index.js

```diff
@@ -1,29 +1,60 @@
 'use strict';
 
-function createError(err, code, props) {
-    var key;
+function assign(obj, props) {
+    let key;
+    let definition;
+
+    for (key in props) {
+        definition = Object.getOwnPropertyDescriptor(obj, key);
+
+        if (definition && !definition.writable) {
+            throw new TypeError(`Cannot assign to read only property '${key}' of object '${obj}'`);
+        }
+
+        obj[key] = props[key];
+    }
+
+    return obj;
+}
 
+function createError(err, code, props) {
     if (!(err instanceof Error)) {
         throw new TypeError('Please pass an Error to err-code');
     }
 
+    if (!props) {
+        props = {};
+    }
+
     if (typeof code === 'object') {
         props = code;
-    } else if (code != null) {
-        try {
-            err.code = code
-        } catch (err) { }
+        code = undefined;
     }
 
-    if (props) {
-        for (key in props) {
-            try {
-                err[key] = props[key]
-            } catch (err) { }
-        }
+    if (code) {
+        props.code = code;
     }
 
-    return err;
+    try {
+        return assign(err, props);
+    } catch (_) {
+        props.message = err.message;
+        props.stack = err.stack;
+
+        const ErrClass = function () {};
+
+        ErrClass.prototype = Object.create(Object.getPrototypeOf(err));
+
+        const newErr = new ErrClass();
+
+        for (const key in props) {
+            Object.defineProperty(newErr, key, {
```

Comment:
Why defineProperty and not a simple assignment like we are doing above?

Reply:
Because if the setter throws, simple assignment fails.  `Object.defineProperty` doesn't invoke the setter.

Reply:
Alright, but perhaps make the props enumerable?

Reply:
Done!